### PR TITLE
Promtool Update: JUnit-Format XML Test Result Support

### PR DESCRIPTION
**Pull request by kushalShukla-web**
Recreated PR for converting junit compataible out put for promtool . 
  solves this #8860

### CODE REVIEW

**Comment by kushalShukla-web** [comment]
Hey @krajorama ,

This PR is to inform you and confirm whether I am heading in the right direction. Please let me know if everything looks good or if any changes are needed.

**Comment by kushalShukla-web** [comment]
Signed-off-by: Kushal Shukla [kushalshukla110@gmail.com]

**Comment by aknuds1** [comment]
Could you fix the failing tests?

**Comment by kushalShukla-web** [comment]
sure @aknuds1

**Comment by kushalShukla-web** [comment]
@aknuds1  here i dont understand why am i getting error like this `Checking ./testdata/rules-bad.yml
  FAILED:
./testdata/rules-bad.yml: group "alerts", rule 1, "InstanceDown": annotation "summary": template: __alert_InstanceDown:1: undefined variable "$label"
Checking ./testdata/prometheus-rules.lint.yml` . i did not even change the prometheus-rules.lint.yml file but it says that label variable is not define .

File: util/junitxml/junitxml.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
+// Copyright 2024 Kushal Shukla
```
Comment (krajorama) [inline util/junitxml/junitxml.go:1]: Please use
```suggestion
// Copyright 2024 The Prometheus Authors
```

Note: using individual name was the reason we rejected the original PR.

File: util/junitxml/junitxml_test.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
+// Copyright 2024 Kushal Shukla
```
Comment (krajorama) [inline util/junitxml/junitxml_test.go:1]:
```suggestion
// Copyright 2024 The Prometheus Authors
```
Note: using individual name was the reason we rejected the original PR.

File: cmd/promtool/unittest.go [hunk 3]
```diff
@@ -18,19 +18,20 @@ import (
...
-	"time"
-
 	"github.com/go-kit/log"
 	"github.com/google/go-cmp/cmp"
 	"github.com/grafana/regexp"
 	"github.com/nsf/jsondiff"
 	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/util/junitxml"
 	"gopkg.in/yaml.v2"
+	"io"
```
Comment (krajorama) [inline cmd/promtool/unittest.go:28]: Linter complains about the ordering of imports, see https://github.com/prometheus/prometheus/actions/runs/9920970887/job/27408406132?pr=14461

File: cmd/promtool/unittest.go [hunk 4]
```diff
@@ -62,27 +68,31 @@ func RulesUnitTest(queryOpts promqltest.LazyLoaderOpts, runStrings []string, dif
 		} else {
 			fmt.Println("  SUCCESS")
 		}
-		fmt.Println()
```
Comment (krajorama) [inline cmd/promtool/unittest.go:65]: I think this `fmt.Println()` can stay.

File: util/junitxml/junitxml.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+type TestSuite struct {
+	Name         string      `xml:"name,attr"`
+	TestCount    int         `xml:"tests,attr"`
+	FailureCount int         `xml:"failures,attr"`
+	ErrorCount   int         `xml:"errors,attr"`
+	SkippedCount int         `xml:"skipped,attr"`
+	Timestamp    string      `xml:"timestamp,attr"`
+	Cases        []*TestCase `xml:"testcase"`
+}
```
Comment (krajorama) [inline util/junitxml/junitxml.go:34]: Let's space out the functions, types a little with an empty line between. Linter complains about it, see https://github.com/prometheus/prometheus/actions/runs/9920970887/job/27408406132?pr=14461. Also for the rest of the file.

```suggestion
}

```

File: cmd/promtool/unittest_test.go [hunk 6]
```diff
@@ -14,11 +14,12 @@
...
-	"testing"
-
-	"github.com/stretchr/testify/require"
-
+	"bytes"
+	"fmt"
+	"github.com/grafana/regexp"
 	"github.com/prometheus/prometheus/promql/promqltest"
+	"github.com/stretchr/testify/require"
+	"testing"
```
Comment (krajorama) [inline cmd/promtool/unittest_test.go:22]: Linter complains about the ordering of imports, see https://github.com/prometheus/prometheus/actions/runs/9920970887/job/27408406132?pr=14461

File: cmd/promtool/unittest.go [hunk 7]
```diff
@@ -91,7 +101,8 @@ func ruleUnitTest(filename string, queryOpts promqltest.LazyLoaderOpts, run *reg
 	}
 
 	evalInterval := time.Duration(unitTestInp.EvaluationInterval)
-
+	currentTime := time.Now().Format("2006-01-02T15:04:05")
+	ts.Settime(currentTime)
 	// Giving number for groups mentioned in the file for ordering.
 	// Lower number group should be evaluated before higher number group.
 	groupOrderMap := make(map[string]int)
```
Comment (krajorama) [inline cmd/promtool/unittest.go:108]: There's a new test here on line 110 compared to the original PR. I think we should create a `ts.Case("unittest_check_ordering")` and record `Fail` when it returns error.

File: util/junitxml/junitxml_test.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+import (
+	"bytes"
+	"errors"
+	"testing"
+
+	"github.com/grafana/regexp"
+)
+
+func TestJunitOutput(t *testing.T) {
```
Comment (krajorama) [inline util/junitxml/junitxml_test.go:24]: nit: this is pretty weird to test XML validity with regexp. I would have expected either a direct compare to a reference string or using encoding/xml to parse and validate.

File: cmd/promtool/unittest_test.go [hunk 9]
```diff
@@ -125,15 +126,54 @@ func TestRulesUnitTest(t *testing.T) {
...
+		}
 		t.Run(tt.name, func(t *testing.T) {
 			if got := RulesUnitTest(tt.queryOpts, nil, false, tt.args.files...); got != tt.want {
 				t.Errorf("RulesUnitTest() = %v, want %v", got, tt.want)
 			}
 		})
 	}
-}
 
+	reTop := regexp.MustCompile(`(?s)^<testsuites\W.*</testsuites>$`)
```
Comment (krajorama) [inline cmd/promtool/unittest_test.go:142]: nit: this is duplicated from  junitxml_test.go , couldn't we just use encoding/xml to parse the output?

File: cmd/promtool/unittest_test.go [hunk 10]
```diff
@@ -125,15 +126,54 @@ func TestRulesUnitTest(t *testing.T) {
 			want: 0,
 		},
 	}
+	reuseFiles := []string{}
+	reuseCount := [2]int{}
```
Comment (krajorama) [inline cmd/promtool/unittest_test.go:130]: this doesn't seem to be initialized later around line 133, like in the original PR

Comment (krajorama) [reply cmd/promtool/unittest_test.go:130]: that's causing the failed test probably

**Comment by krajorama** [comment]
> @aknuds1 here i dont understand why am i getting error like this `Checking ./testdata/rules-bad.yml FAILED: ./testdata/rules-bad.yml: group "alerts", rule 1, "InstanceDown": annotation "summary": template: __alert_InstanceDown:1: undefined variable "$label" Checking ./testdata/prometheus-rules.lint.yml` . i did not even change the prometheus-rules.lint.yml file but it says that label variable is not define .

I think those printouts about the rules are expected output from the code, the actual issue is 
```
  unittest_test.go:164: JUnit output had 2 passes; expected 0
        unittest_test.go:167: JUnit output had 7 failures; expected 0
```
I did comment on what I think the issue is: https://github.com/prometheus/prometheus/pull/14461#discussion_r1678855024

File: util/junitxml/junitxml_test.go (hunk 8 above)
Comment (kushalShukla-web) [reply util/junitxml/junitxml_test.go:24]: what im thinking like , can't we convert the yaml file data into struct data and i already have struct data as per encoding/xml package and then we can compare them .

**Comment by kushalShukla-web** [comment]
Everything is fine, @krajorama. Just need to remove some print statements and ensure DCO sign-off.

**Comment by kushalShukla-web** [comment]
ig that last error is related to native histogram part and nothing to do with this issue .

**Pull request closed without merging**